consensus/hotstuff/safetyrules/safety_rules.go

```diff
@@ -93,16 +105,76 @@ func (v *SafetyRules) ProduceVote(proposal *model.Proposal, curView uint64) (*mo
 //  * (nil, model.NoTimeoutError): If the safety module decides that it is not safe to timeout under current conditions.
 //    This is a sentinel error and _expected_ during normal operation.
 // All other errors are unexpected and potential symptoms of uncovered edge cases or corrupted internal state (fatal).
-func (v *SafetyRules) ProduceTimeout(curView uint64, highestQC *flow.QuorumCertificate, highestTC *flow.TimeoutCertificate) (*model.TimeoutObject, error) {
-	panic("to be implemented")
+func (r *SafetyRules) ProduceTimeout(curView uint64, highestQC *flow.QuorumCertificate, highestTC *flow.TimeoutCertificate) (*model.TimeoutObject, error) {
+	lastTimeout := r.safetyData.LastTimeout
+	if lastTimeout != nil && lastTimeout.View == curView {
+		return lastTimeout, nil
+	}
+
+	if !r.IsSafeToTimeout(curView, highestQC, highestTC) {
+		return nil, model.NoTimeoutError{Msg: "not safe to time out under current conditions"}
+	}
+
+	timeout, err := r.signer.CreateTimeout(curView, highestQC, highestTC)
+	if err != nil {
+		return nil, fmt.Errorf("could not timeout at view %d: %w", curView, err)
+	}
+
+	r.safetyData.HighestAcknowledgedView = curView
+	r.safetyData.LastTimeout = timeout
+
+	err = r.persist.PutSafetyData(r.safetyData)
+	if err != nil {
+		return nil, fmt.Errorf("could not persist safety data: %w", err)
+	}
+
+	return timeout, nil
 }
 
 // IsSafeToVote checks if this proposal is valid in terms of voting rules, if voting for this proposal won't break safety rules.
-func (v *SafetyRules) IsSafeToVote(proposal *model.Proposal) bool {
-	panic("to be implemented")
+func (r *SafetyRules) IsSafeToVote(proposal *model.Proposal) error {
+	blockView := proposal.Block.View
+	qcView := proposal.Block.QC.View
+
+	// block's view must be larger than the view of the included QC
+	if blockView <= qcView {
+		return fmt.Errorf("block's view %d must be larger than the view of the included QC %d", blockView, qcView)
+	}
+
+	// block's view must be greater than the view that we have voted for
+	if proposal.Block.View <= r.safetyData.HighestAcknowledgedView {
+		return model.NoVoteError{Msg: "not safe to vote, we have already voted for this view"}
+	}
+
+	// This check satisfies voting rule 1 and 2a.
+	if blockView == qcView+1 {
+		return nil
+	}
+
+	// This check satisfies voting rule 1 and 2b.
+	lastViewTC := proposal.LastViewTC
+	if lastViewTC != nil {
```

Comment:
I think it is worth factoring this into a `safeToExtend` function, like in the formal algorithm.
```suggestion
	return isSafeToExtend(blockView, qcView, lastViewTC)
```

What do you think about handling the failure cases first to reduce nesting and move the errors closer to the conditions that cause them? eg:
```
if blockView != lastViewTC.View+1 {
  // error
}

if qcView < lastViewTC.TOHighestQC.View {
  // error
}

// We satisfy voting rule 1 and 2b ...
return nil